src/server/game/Craft/Crafting.cpp

```diff
@@ -0,0 +1,243 @@
+
+    uint32 qualityTier = sDB2Manager.GetCraftingQualityTierByDifficultyPercent(_craftingData->CraftingDifficultyID, difficultyPercentMatched);
+
+    uint32 craftedItemId = GetCraftedItemIdForQuality(qualityTier);
+
+    // If we didn't found anything, we have a problem. Return and don't take reagent from player.
+    if (!craftedItemId)
+        return SpellCastResult::SPELL_FAILED_ERROR;
+
+    ItemContext context = _spell->GetSpellInfo()->HasAttribute(SPELL_ATTR0_IS_TRADESKILL) ? ItemContext::Trade_Skill : ItemContext::NONE;
+
+    std::vector<int32> bonusLists;
+    for (int32 bonusTreeID : bonusTreeIds)
+    {
+        std::vector<int32> treeBonusList = ItemBonusMgr::GetBonusListsForItem(craftedItemId, { context, {}, {}, bonusTreeID, qualityTier });
+        bonusLists.insert(std::end(bonusLists), std::begin(treeBonusList), std::end(treeBonusList));
+    }
+
+    if (Item* createdItem = _spell->DoCreateItem(craftedItemId, context, &bonusLists, false))
+    {
+        InitCraftingStatModifier(createdItem);
+
+        createdItem->AddToWorld();
+        createdItem->SendUpdateToPlayer(_player);
+    }
+
+    // Item has been crafted, remove reagent from player
+    for (auto const& [itemId, itemCount] : usedReagentsToDestroy)
+        _player->DestroyItemCount(itemId, itemCount, true);
+
+    return SpellCastResult::SPELL_CAST_OK;
+}
+
+uint32 Crafting::GetSkillLevelForCraft()
+{
+    uint32 skillId = GetSkillIdForSpell(_spell->GetSpellInfo()->Id);
+    uint32 skillValue = _player->GetSkillValue(skillId);
+
+    uint32 totalReagentWeigth = CalculateTotalReagentWeights();
+
+    CraftingDifficultyEntry const* craftingDifficultyEntry = sCraftingDifficultyStore.LookupEntry(_craftingData->CraftingDifficultyID);
+    if (!craftingDifficultyEntry)
+        return skillValue;
+
+    float maximumReagentBonus = round(float(_craftingData->CraftingDifficulty) * float(craftingDifficultyEntry->CraftSkillBonusPercent) / 100.f);
+
+    for (ModifiedCraftingSpellSlotEntry const* spellSlot : sDB2Manager.GetMCRSpellSlotBySpell(_spell->GetSpellInfo()->Id))
+    {
+        for (ModifiedCraftingCategoryEntry const* category : sDB2Manager.GetMCRCategoryByReagentSlot(spellSlot->ModifiedCraftingReagentSlotID))
+        {
+            auto providedReagentItr = _spell->m_targets.ModifiedCraftingReagents.find(spellSlot->Slot);
+            if (providedReagentItr != _spell->m_targets.ModifiedCraftingReagents.end())
+            {
+                for (WorldPackets::Spells::SpellCraftingReagent const& providedReagentForSlot : providedReagentItr->second)
+                {
+                    if (CraftingReagentQualityEntry const* reagentQuality = sDB2Manager.GetCraftingReagentQualityByItem(providedReagentForSlot.ItemID))
+                    {
+                        float reagentWeight = float(category->MatQualityWeight) * float(providedReagentForSlot.Quantity) / 2.0f * float(reagentQuality->OrderIndex);
```

Comment:
why is it divided by 2.f? no documentation as to why, and I really recommend encapsulating it in a function like CalculateReagentWeight()